lib/ansible/modules/storage/netapp/na_ontap_broadcast_domain_ports.py

```diff
@@ -0,0 +1,237 @@
+
+
+class NetAppOntapBroadcastDomainPorts(object):
+    """
+        Create and Destroys Broadcast Domain Ports
+    """
+    def __init__(self):
+        """
+            Initialize the Ontap Net Route class
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present', 'absent'], default='present'),
+            vserver=dict(required=False, type='str'),
+            broadcast_domain=dict(required=True, type='str'),
+            ipspace=dict(required=False, type='str', default=None),
+            ports=dict(required=True, type='list'),
+        ))
+
+        self.module = AnsibleModule(
+            argument_spec=self.argument_spec,
+            supports_check_mode=True
+        )
+
+        parameters = self.module.params
+
+        # set up state variables
+        self.state = parameters['state']
+        self.vserver = parameters['vserver']
+        self.broadcast_domain = parameters['broadcast_domain']
+        self.ipspace = parameters['ipspace']
+        self.ports = parameters['ports']
+
+        if HAS_NETAPP_LIB is False:
+            self.module.fail_json(msg="the python NetApp-Lib module is required")
+        else:
+            self.server = netapp_utils.setup_na_ontap_zapi(module=self.module)
+        return
+
+    def get_broadcast_domain_ports(self):
+        """
+        Return details about the broadcast domain ports
+        :param:
+            name : broadcast domain name
+        :return: Details about the broadcast domain. None if not found.
+        :rtype: dict
+        """
+        domain_get_iter = netapp_utils.zapi.NaElement('net-port-broadcast-domain-get-iter')
+        broadcast_domain_info = netapp_utils.zapi.NaElement('net-port-broadcast-domain-info')
+        broadcast_domain_info.add_new_child('broadcast-domain', self.broadcast_domain)
+        query = netapp_utils.zapi.NaElement('query')
+        query.add_child_elem(broadcast_domain_info)
+        domain_get_iter.add_child_elem(query)
+        result = self.server.invoke_successfully(domain_get_iter, True)
+        domain_exists = None
+        # check if broadcast domain exists
+        if result.get_child_by_name('num-records') and \
+                int(result.get_child_content('num-records')) == 1:
+            domain_info = result.get_child_by_name('attributes-list').get_child_by_name('net-port-broadcast-domain-info')
+            domain_name = domain_info.get_child_content('broadcast-domain')
+            domain_ports = domain_info.get_child_content('port-info')
+            domain_exists = {
+                'domain-name': domain_name,
+                'ports': domain_ports
+            }
+        return domain_exists
+
+    def create_broadcast_domain_ports(self):
+        """
+        Creates new broadcast domain ports
+        """
+        domain_obj = netapp_utils.zapi.NaElement('net-port-broadcast-domain-add-ports')
```

Comment:
I'm also making the rash assumption that the underlying API is itself idempotent (eg that it will succeed if you ask it to set a state that it's already in)- if not, this is a much bigger problem.

Reply:
Some API are some are not. In this case it is not.

So what we've done lower down in the function is have a try/except, if it passes without an error return true (which is the changed state). if we get an exception with error code 18605, then we know the port is already being used so we returned False for the change state.

        try:
            self.server.invoke_successfully(domain_obj, True)
            return True
        except netapp_utils.zapi.NaApiError as error:
            # Error 18605 denotes port already being used.
            if to_native(error.code) == "18605":
                return False